cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
 	wg.Wait()
 	return resultingKubeConfig
 }
+
+// certChain is an array of *x509.Certificate with a helper to plug into the tls VerifyPeerCertificate hook
+type certChain []*x509.Certificate
+
+// saveCertificateChain parses and saves the array of DER-encoded certificates into a certChain
+func (c *certChain) saveCertificateChain(rawCerts [][]byte, _ [][]*x509.Certificate) error {
```

Comment:
Yeah, validating the entire initial connection gives better trust in the other parts of the `cluster-info` ConfigMap since we use it for more than just the cluster CA. I think we could also accomplish this by making two requests if we want to go that route.

We need the full chain since the server might have some intermediate CAs included that we need to chain through to get from the root to the leaf cert.